### ISSUE TITLE
np.take and np.einsum aren't fused properly

### ISSUE THREAD

**Issue by AranKomat** [google/jax#3206]
I'm trying to translate [Product Key Memory ](https://arxiv.org/abs/1907.05242) in PyTorch into JAX, and this requires the translation of [nn.EmbeddingBag](https://pytorch.org/docs/master/generated/torch.nn.EmbeddingBag.html) with per_sample_weights, as I haven't found any counterpart in JAX (but if you know, please let me know). For this, I wrote scatter_weighted_sum, the weighted sum version of scatter_add, in hope that it'll be efficient with fusing. However, jit didn't fuse np.take, reshape and np.einsum properly, which resulted in a huge intermediate object. Since #1979 concluded this sort of ops will be fused on GPU, I was wondering what is causing this problem. If by any chance this isn't supported on GPU, should this work on TPU? I'm using JAX ver. 0.1.67 with various Colab GPUs.   

````python
hidden_dim = 512
n_keys = 512
batch = 2 ** 15
knn = 32
heads = 4
key = random.PRNGKey(0)
values = random.normal(key, (n_keys ** 2, hidden_dim))         
indices = random.randint(key, (batch*heads, knn), 0, n_keys ** 2)     
weights = random.normal(key, (batch*heads, knn))     
@jit
def scatter_weighted_sum(inputs, indices, weights): 
    num_bags = weights.shape[-1]
    dim = inputs.shape[-1]
    indices = indices.reshape(-1)
    tmp = inputs.take(indices, axis=0).reshape(-1, num_bags, dim)
    return np.einsum('ind, in -> id', tmp, weights)

````

**Comment by hawkinsp** [comment]
I agree, XLA is not fusing the gather with the reduction, which would avoid materializing the intermediate value, and I don't see a good reason for that. I filed a couple of bugs for the XLA teams.

A possible workaround would be to use `lax.map` to compute the elements in chunks.

**Comment by bhack** [comment]
We are having the same problem in TF-addons (and so a green light for a custom-op PR). https://github.com/tensorflow/addons/issues/2201

And in XLA/Pytorch:
https://github.com/pytorch/xla/issues/2403

**Comment by bhack** [comment]
@hawkinsp Are these XLA bugs/tickets public on github or internal?

**Comment by hawkinsp** [comment]
It's currently an internal bug (b/157497992). It's worth noting that I raised this issue in particular with respect to GPUs; the issues might be different on CPU, GPU, or TPU. I'll ping it, although there are no guarantees they will have time...

**Comment by bhack** [comment]
@hawkinsp Do you think that you can open one at https://github.com/tensorflow/mlir-hlo so that It would be easier to track?

**Comment by hawkinsp** [comment]
Given that in this context we are talking about XLA classic not any of the MLIR work, the best place for a public XLA issue is the TF Github tracker. Feel free to file one if you like, although I think this issue covers it just as well.

**Comment by bhack** [comment]
In core TF we have already two users tickets that depend on this XLA missing fusion. I don't know if the team plans to cover that on MLIR-HLO or it is still plausible to expect something for classic XLA:
https://github.com/tensorflow/tensorflow/issues/32675
https://github.com/tensorflow/tensorflow/issues/43784

**Comment by shz0116** [comment]
Using embeddingbag instead of mebedding+allreduce will have a "huge" impact not only on performance but also on memory footprint for large number of indices. By computing the "sum" operation on the fly will save great amount of memory. We do not need to store all the table lookup results. A higher priority should be given to this issue.
